### DISCUSSION TITLE
Visual Studio Code not using project environment

### DISCUSSION THREAD

**Discussion by czaefferer** [pingbird/puro#42] [Q&A]
I have configured "stable" as the global environment, and created an environment called "v3_7_6" for Flutter 3.7.6.

For an app still using Flutter 3.7, I configured puro to use this environment. If I run "puro flutter run" from the terminal in the project directory, it runs the app with the 3.7 environment.

However, if I try to start the app from Visual Studio Code, it will use the stable environment, instead of the configured one.
The settings.json file created by puro looks like this:

```
{
    "dart.flutterSdkPath": "/home/<user>/.puro/envs/v3_7_6/flutter",
    "dart.sdkPath": "/home/<user>/.puro/envs/v3_7_6/flutter/bin/cache/dart-sdk"
}
```

What do I need to do?

**Comment by tksoh** [comment]
what is the output of `puro ls` in the project directory?

**Comment by tksoh** [comment]
And how do you know VSCode is using stable instead of v3_7_6?

**Comment by czaefferer** [comment]
this is the output of `puro ls`:
![image](https://github.com/pingbird/puro/assets/47598945/34993ed9-57be-4b65-bdd1-a050eefc81bd)

If I add a parameter for verbose-output to the launch configuration, I can see the path to the Flutter SDK used (also the build doesn't suceed with a newer SDK).

It looks to me like my VSCode can handle only one SDK for the whole workspace, not on a per-project base? WIth the "Auto-detect SDK location" it will use an SDK path from a (random?) project in the workspace, and then uses that for all projects that it launches or I edit.

For now I have removed all the settings related to the Flutter SDK, and manually added the SDK paths (where puro stores the SDKs) to "dart.flutterSdkPaths", and switch the SDK for the whole workspace manually whenever needed.

**Comment by tksoh** [comment]
@czaefferer do you mind sharing the launch configuration, and the build output that shows the flutter SDK info? Maybe I can try to recreate on my system to see. Also. what platform (Windows 11, etc) are you using to develop flutter, and also VS Code versions?

**Comment by tksoh** [comment]
@czaefferer one other thing you might also want to try, if you haven't already, is to enable the "Add Sdk To Terminal Path" option in VSCode settings, then run `puro use` to change the env, then open a new terminal in VS Code (so it can pick up the new path) and use `which flutter` see if it's the flutter from the puro env you just set. You can also use Ctrl-Alt-P to add the path, but it doesn't work for me all the time somehow.

![Screenshot 2023-09-23 194157](https://github.com/pingbird/puro/assets/494222/edc5c19a-1ce0-4623-9227-226676e6bb0b)

BTW, in theory every time you change the env, VS Code should prompt to run `pub get` if the version change is significant enough to affect the pubspecs. If you are not getting the prompt, then perhaps there's some issue with your VS Code setup.

![Screenshot 2023-09-23 195331](https://github.com/pingbird/puro/assets/494222/2698d91b-d9d2-4b89-b5ad-2a27ddbf0821)

**Comment by pingbird** [comment]
@czaefferer Make sure the `.vscode/settings.json` is at the root of your project and not a sub-folder, the Flutter plugin should automatically use what is configured there.

**Comment by czaefferer** [comment]
@tksoh I've got "dart.addSdkToTerminalPath" activated, when I switch the SDK manually this change is reflected in the CLI, so that works from what I can tell. And also VSCode does indeed prompt me to update dependencies whenever I switch the SDK. It's just that VSCode doesn't switch the SDK by itself. To make it easier to reproduce, I've created a new workspace and two new projects for testing. I'm using Ubuntu 22.04.3 LTS and VSCode 1.82.2. These are the steps I've done:

- Create a new workspace
- remove "dart.flutterSdkPaths" and "dart.sdkPath" from settings, keep "dart.addSdkToTerminalPath": true from user-settings
- From commandline (within VSCode or external): "which flutter"
> /home/cza/.puro/envs/default/flutter/bin/flutter

---

- flutter create --platforms android puro_test_3_7
- add project to workspace in VSCode
- puro use v3_7_6
> [✓] Switched to environment `v3_7_6`
- a file .vscode/settings.json was created in the project with content:
> {
>     "dart.flutterSdkPath": "/home/cza/.puro/envs/v3_7_6/flutter",
>     "dart.sdkPath": "/home/cza/.puro/envs/v3_7_6/flutter/bin/cache/dart-sdk"
> }
- setting sdk in pubspec.yaml to "2.19.3"
- VSCode is still using stable for Flutter (3.13.3), and I can not switch the SDK manually. VSCode only seems to know the default SDK, the SDK path from the projects-settings file seems to be completly ignored. I can now run the project neither from VSCode nor with "flutter run" from the CLI, since they will both complain about the wrong SDK:
> The current Dart SDK version is 3.1.1.
> Because puro_test_3_7 requires SDK version 2.19.3, version solving failed.
- however "puro flutter --version" will show the correct version:
> Flutter 3.7.6 • channel stable • https://github.com/flutter/flutter.git
> Framework • revision 12cb4eb7a0 (7 months ago) • 2023-03-01 10:29:26 -0800
> Engine • revision ada363ee93
> Tools • Dart 2.19.3 • DevTools 2.20.1
- and "puro flutter run" will use the correct sdk and runs the app

---

- now I create a second project: "flutter create --platforms android puro_test_3_13"
- add project to workspace in VSCode
- use the 3.13.0 SDK: "puro use v3_13_0"
- a file .vscode/settings.json was created in the project with content:
> {
>     "dart.flutterSdkPath": "/home/cza/.puro/envs/v3_13_0/flutter",
>     "dart.sdkPath": "/home/cza/.puro/envs/v3_13_0/flutter/bin/cache/dart-sdk"
> }
- set the SDK version in pubspec.yaml to "3.1.0"
- basically same result as with the 3.7 app, new SDK is not known to VSCode, app cannot be run other than "puro flutter run"

---

- My current workaround is to add this to my user-settings:
> "dart.flutterSdkPaths": [
>     "/home/cza/.puro/envs/v3_7_6/flutter",
>     "/home/cza/.puro/envs/v3_13_0/flutter",
>     "/home/cza/.puro/envs/stable/flutter"
> ]
- now VSCode knows these three SDKs and allows me to manually change the SDK. "Auto-detect SDK location" chosses the 3.7 environment for the whole workspace, If I wanted to run the 3.13 app I'd have to manually switch the SDK
![Screenshot from 2023-09-25 09-48-27](https://github.com/pingbird/puro/assets/47598945/820e2f1d-d0cc-441c-8611-faad10189b3a)
- The chosen Flutter SDK is now also added to the CLI within VSCode, but for all projects, even the 3.13 one: "which flutter"
> /home/cza/.puro/envs/v3_7_6/flutter/bin//flutter

---

The final folder-structure (after some cleanup) looks like this:
![Screenshot from 2023-09-25 09-58-29](https://github.com/pingbird/puro/assets/47598945/523f4a82-7e39-47e2-b618-a2b002e20e05)

**Comment by tksoh** [comment]
@czaefferer 

> I've got "dart.addSdkToTerminalPath" activated, when I switch the SDK manually this change is reflected in the CLI, so that works from what I can tell. 

By "switch the SDK manually", do you mean by running `puro use` in the terminal of the project in vs code?

**Comment by czaefferer** [comment]
@tksoh No, I've got "Flutter SDK" in the status bar, where I can see the currently active SDK when hovering over it, and select an SDK from "dart.flutterSdkPaths" when clicking:
![image](https://github.com/pingbird/puro/assets/47598945/422260c5-769d-4e75-872d-80575a51cdf1)

If nothing is configured in "dart.flutterSdkPaths", it will show the default Flutter SDK (from /home/cza/.puro/envs/default/flutter/bin/flutter) and not allow me to change it, even with the .vscode/settings.json files in the projects.

**Comment by tksoh** [comment]
> And also VSCode does indeed prompt me to update dependencies whenever I switch the SDK.

And this happened when you switch via 'puro use' to switch, as well as manually changing the path via vs code?

BTW, thanks for capturing such details in your earlier post. It helps a lot.

**Comment by czaefferer** [comment]
> And this happened when you switch via 'puro use' to switch, as well as manually changing the path via vs code?

No, only when I switch manually. I can see the explorer in VSCode update when the settings-files are created, but no prompt to update dependencies.

**Comment by tksoh** [comment]
> No, only when I switch manually. I can see the explorer in VSCode update when the settings-files are created, but no prompt to update dependencies.

Hmm... seems to me that your VSCode might be ignoring your workspace settings (.vscode/settings.json) somehow. I did a quick google on the topic, but haven't found much discussion on this issue.

Perhaps you can experiment with the workspace settings for your projects, to confirm if VSCode would pick them up as intended. If indeed they don't, then this might be something to ask on the VSCode forums.

**Comment by czaefferer** [comment]
I've played around settings different values for "dart.lineLength" in the projects-settings and user-settings, and the project-settings get ignored on my main machine. So something is indeed wrong there.

For an additional test I tried all the setup from this page on a different machine that I rarely use and have not customized other than installing the dart and flutter extensions, and there the "dart.lineLength" from the project-settings is picked up. The dart/flutter SDK paths however only partially: Initially when I had both the 3.7 and 3.13 projects in my workspace it used 3.7 as the SDK. When I removed the 3.7 project from the workspace it changed to the 3.13 SDK. So the path from both project-setting-files was picked up at some point.
However after that, whatever I did, the SDK would not change anymore and stay with 3.13, even if only the 3.7 project is in the workspace.

And just for clarification what the expected behaviour is: If I have both a projekt using the 3.7 SDK and one with the 3.13 SDK in my workspace, whenever I open a dart file from the other project, VSCode should change the SDK used, and that change should be reflected in the statusbar (3.7.6 in the screenshot)?
![image](https://github.com/pingbird/puro/assets/47598945/95592d00-0cc5-47fa-92de-d7ed94159c78)

**Comment by tksoh** [comment]
> And just for clarification what the expected behaviour is: If I have both a projekt using the 3.7 SDK and one with the 3.13 SDK in my workspace, whenever I open a dart file from the other project, VSCode should change the SDK used, and that change should be reflected in the statusbar (3.7.6 in the screenshot)?

I don't quite understanding what you meant by "whenever I open a dart file from the other project". I usually have one VSCode window for each flutter/dart project, except on rare occasions I would open some file outside those projects, such as ~/.bashrc, etc.

BTW, how do you get the Flutter SDK version to show up on the status bar. I can't quite find the way to get the VSCode on my Windows 11 PC to display that. Not sure if that might tell us there's some difference between VSCode on Windows vs Linux.

**Comment by czaefferer** [comment]
> I don't quite understanding what you meant by "whenever I open a dart file from the other project". I usually have one VSCode window for each flutter/dart project, except on rare occasions I would open some file outside those projects, such as ~/.bashrc, etc.

Okay, then you're not using a workspace ;) 
Open a new (empty) VSCode window and via `File` -> `Add Folder to Workspace` add two projects with different SDK requirements to the workspace. Now you've got multiple projects open in the same VSCode, but only one Flutter SDK. This is how I'm usually using VSCode, since I have multiple apps and libraries to work on during the day.

> BTW, how do you get the Flutter SDK version to show up on the status bar. I can't quite find the way to get the VSCode on my Windows 11 PC to display that. Not sure if that might tell us there's some difference between VSCode on Windows vs Linux.

Open a dart file, and hover over the `{}` next to `Dart` in the status bar. There you can pin the Flutter SDK to the status bar:
![image](https://github.com/pingbird/puro/assets/47598945/ee3c364e-76bb-4603-b665-f5e81fc99619)

**Comment by tksoh** [comment]
@czaefferer 

> Okay, then you're not using a workspace ;)

Indeed I haven't 😅 So I gave it a tried, and here's what I found...

 If I open my 'test' project as I normally do without workspace,  VSCode picks up the flutter version 3.7.12 set by `puro use legacy`. However, if I add the same 'test' project to a workspace, VSCode always pick up the default flutter v3.13.5 (as found on PATH I presumed), ignoring `puro use legacy`. 

Here's the output of `puro ls` in 'test' project:

```
$ puro ls
[i] Environments:
    ~ stable (stable / 3.13.5 / 12fccda598)
      beta   (not installed)
      master (not installed)
    * legacy (stable / 3.7.12 / 4d9e56e694)
    
    Use `puro create <name>` to create an environment, or `puro use <name>` to switch
```

Perhaps VS Code ignores `.vscode/settings.json` while in workspace mode, since the projects added to the workspace are sharing the flutter SDK like you mentioned? What would happen if you try open your projects in non-workspace mode, to see if VSCode picks up the  `.vscode/settings.json`?

@pingbird any thought?

**Comment by tksoh** [comment]
@czaefferer 

More discovery... According to VSCode doc, the settings for a multi-root workspace are actually stored inside the `.code-workspace` file. Here's my copy:

```
{
	"folders": [
		{
			"path": "test"
		}
	],
	"settings": {
		"dart.flutterSdkPath": "~/.puro/envs/stable/flutter"
	}
}
```

By editing the content of `dart.flutterSdkPath` in this settings object of .code.workspace file, VSCode is able to pick up the flutter version accordingly. 

I guess it now comes down to how puro would support multi-root workspace. Until then, you can manually update the path to point to puro env of your choice. It's not ideal, but it's still better than add the paths to user settings manually.

**Comment by tksoh** [comment]
BTW, the doc also mentions that "_Settings configured per folder will override settings defined in the .code-workspace_". Although in my experiment,  VSCode does NOT pick up the flutter path in `test/.vscode/settings.json` file. I'm not sure whether it's talking about some other settings file. Maybe there's something I'm missing here.

![Screenshot 2023-10-07 105059](https://github.com/pingbird/puro/assets/494222/bd0ecc2c-2ed8-41cd-9289-f8fa99f91319)
